### PR TITLE
DEV-4006 | Unbreak dashboard (from login to pages list view) for superusers and hub admins

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### What's this PR do?

This PR attempts to make the pages list view of the org dashboard useable again by:

- Solving for n+1 queries on api/v1/users 
- Solving for inefficient slow queries (but apparently not n+1) due to double serialization of some data in api/v1/pages

It significantly reduces the payload size returned by /api/v1/users and /api/v1/pages — especially for hub admins and superusers.

Along the way:

- It simplifies the serialized representation of users returned by `api/v1/token` and `api/v1/users` such that the two endpoints use a shared schema for representing user.
- Renames and otherwise cleans up existing serializers used for representing pages and users (and organizations and revenue programs therein).
- Refactors apps/users/tests/test_serializers.py to only use pytest. In some cases, when given chance between redundantly testing serializer code in serialier unit tests that is also tested in view layer, I opted to remove the serializer tests as they were providing no additional value, and tended to look more like tests of implementation details.  
- Refactors apps/users/tests/test_views.py to only use pytest. 
- Light changes to some type definitions in spa 

#### Why are we doing this? How does it help us?

Make the org dashboard usable for superusers and hub admins

#### Are there detailed, specific, step-by-step testing instructions in the associated ticket?

Yes, however, the testing in review app can only be vs. regressions. The bugs we've been seeing seem to uniquely happen in prod, so we'll need to test out vs. prod.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No

#### Have automated unit tests been added? If not, why?

Yes

#### How should this change be communicated to end users?

We should let NRH staff who use dashboard as superusers and org admins that they should expect basic usability moving forward.

#### Are there any smells or added technical debt to note?

Not added, but noted: I think it's a code smell that we're maintaining two separate API endpoints that provide potentially divergent truths about the state of a user. I don't think it's duty of this PR to solve for that, but the argument for it seems clear when looking at changes in this PR. [DEV-4031]

#### What are the relevant tickets? Add a link to any relevant ones.

- [DEV-4006](https://news-revenue-hub.atlassian.net/browse/DEV-4006)
- [DEV-4067](https://news-revenue-hub.atlassian.net/browse/DEV-4067)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)? If so, open a ticket and link it to the ticket for this PR and list it here:

No.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

- [DEV-4030](https://news-revenue-hub.atlassian.net/browse/DEV-4030)
- [DEV-4067](https://news-revenue-hub.atlassian.net/browse/DEV-4067)
- [DEV-4082](https://news-revenue-hub.atlassian.net/browse/DEV-4082)

[DEV-4030]: https://news-revenue-hub.atlassian.net/browse/DEV-4030?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DEV-4006]: https://news-revenue-hub.atlassian.net/browse/DEV-4006?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DEV-3909]: https://news-revenue-hub.atlassian.net/browse/DEV-3909?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DEV-4031]: https://news-revenue-hub.atlassian.net/browse/DEV-4031?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/api/serializers.py [hunk 1]
```diff
@@ -3,30 +3,15 @@
...
 # Import error messages to set defaults for fields
 import apps.api.error_messages  # noqa
 from apps.api.tokens import ContributorRefreshToken
-from apps.users.serializers import UserSerializer
 
 
 logger = logging.getLogger(f"{settings.DEFAULT_LOGGER}.{__name__}")
 
 
-class TokenObtainPairCookieSerializer(TokenObtainPairSerializer):
```
Comment (benjaminEwhite) [inline apps/api/serializers.py:17]: This change happens in DEV-3909, which has been reviewed separately.

File: spa/src/components/settings/Organization/Organization.tsx [hunk 2]
```diff
@@ -59,7 +59,7 @@ const Organization = () => {
     [user?.organizations]
   );
   const revenueProgramFromCurrentOrg = useMemo(
-    () => user?.revenue_programs?.filter((rp) => rp.organization.id === currentOrganization?.id),
+    () => user?.revenue_programs?.filter((rp) => rp.organization === currentOrganization?.id),
```
Comment (benjaminEwhite) [inline spa/src/components/settings/Organization/Organization.tsx:62]: We just return org as an ID now, not as a nested object.

File: apps/users/views.py [hunk 3]
```diff
@@ -199,22 +215,14 @@ def get_permissions(self):
...
-        if self.action == "customize_account":
-            permission_classes = [UserOwnsUser, IsAuthenticated, UserIsAllowedToUpdate, UserHasAcceptedTermsOfService]
 
         return [permission() for permission in permission_classes]
 
     def send_verification_email(self, user):
         """Send email to user asking them to click verify their email address link."""
-        if not user.email:
-            logger.warning("Account Verification: No email for user: %s", user.id)
-            return
```
Comment (njharman) [inline apps/users/views.py:217]: Are really email-less users no longer possible?

Sorry, lots of deletes, small changes; it's hard to follow everything.

**Review by njharman — CHANGES_REQUESTED** [review]
Little harder to follow these changes.  comments, change requests in comments.

haven't checked coverage, I have no more time today; will have to do in rereview.

File: apps/organizations/views.py [hunk 4]
```diff
@@ -309,7 +309,7 @@ def handle_stripe_account_link(request, rp_pk):
     This endpoint is designed to support polling from the front end.
     """
     revenue_program = get_object_or_404(RevenueProgram, pk=rp_pk)
-    if not request.user.roleassignment.can_access_rp(revenue_program):
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/views.py:312]: Backed out of this change, which is a nice to have, but need not be in scope.

4d10661c6f7f

Created a follow up ticket to standardize use of user.permitted_revenue_programs, user.permitted_organizations, user.active_flags (and one we should create, user.permitted_styles) across app.

File: apps/pages/serializers.py [hunk 5]
```diff
@@ -475,20 +476,36 @@ def to_representation(self, instance):
...
+
+
 class DonationPageListSerializer(serializers.ModelSerializer):
-    revenue_program = RevenueProgramListInlineSerializer(read_only=True)
+    """Expected usage is representing lists of donation pages in api/v1/pages/
+
+    The primary consumer of this page at time of this comment is the SPA, and specifically
+    the pages list view in the org dashboard.
+
+    Note that at present, pagiation is not enabled for this serializer, and superusers and hub
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/pages/serializers.py]: d3abe335f

File: apps/pages/views.py [hunk 6]
```diff
@@ -220,6 +218,23 @@ def destroy(self, request, pk):
...
+        NB: This method assumes that any queryset filtering vis-a-vis user role and identity has
+        already done upstream in get_queryset.
+
+
+        We prefetch_related on revenue program because that makes the queries required for DonationPageListSerializer
+        to run much faster.
+
+        """
+        serializer = serializers.DonationPageListSerializer
+        # serializer defines its own fields that can be passed to .only() in order to make queries mroe performant
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/pages/views.py]: Code comment removed in 3887ecc

File: apps/users/views.py (hunk 3 above)
Comment (benjaminEwhite) [reply apps/users/views.py:217]: FWIW, the reason this even came in scope was that I had a single line in apps/users/views.py that was untested, and I was tempted to get to 100%.

**Review by nrh-cklimas — CHANGES_REQUESTED** [review]

File: spa/src/hooks/useContributionPage/useContributionPage.types.ts [hunk 7]
```diff
@@ -329,6 +329,9 @@ export interface RevenueProgram {
   tax_id?: string | null;
 }
 
+export interface ContributionPageRevenueProgram extends Omit<RevenueProgram, 'organization'> {
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply spa/src/hooks/useContributionPage/useContributionPage.types.ts]: ccc04b61eedd

File: spa/src/components/settings/Organization/Organization.tsx (hunk 2 above)
Comment (benjaminEwhite) [reply spa/src/components/settings/Organization/Organization.tsx]: That makes sense.

I pushed up 67d2008eb8

@nrh-cklimas can you provide detail on expanded QA scope?

Comment (nrh-cklimas) [reply spa/src/components/settings/Organization/Organization.tsx]: @benjaminEwhite I mean we'd need to test this Organization settings page with org admin and RP admin, which currently isn't part of the testing instructions. It's not a lot, though.

**Review by nrh-cklimas — APPROVED** [review]

File: spa/src/components/settings/Organization/Organization.tsx (hunk 2 above)
Comment (benjaminEwhite) [reply spa/src/components/settings/Organization/Organization.tsx]: Updated testing instructions accordingly

**Review by njharman — APPROVED** [review]

File: pyproject.toml [hunk 8]
```diff
@@ -77,6 +77,7 @@ mailchimp-marketing = "^3.0.80"
 google-cloud-secret-manager = "^2.16.0"
 django-request-id = "^1.0.0"
 pydantic = "^2.2.0"
+dateparser = "^1.1.8"
```
Comment (x110dc) [inline pyproject.toml:80]: It looks like this is only used in one test right? Shouldn't it be a dev dependency then?

**Review by x110dc — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 8 above)
Comment (benjaminEwhite) [reply pyproject.toml]: Took care of this when merging in main here: f26b1bcbb

**Review by x110dc — APPROVED** [review]
